sys/net/gnrc/network_layer/ipv6/ext/frag/gnrc_ipv6_ext_frag.c

```diff
@@ -46,17 +60,256 @@ void gnrc_ipv6_ext_frag_init(void)
+            (snd_buf->pkt->size <= remaining)) {
+            ptr = snd_buf->pkt;
+            snd_buf->pkt = ptr->next;
+            ptr->next = NULL;
+        }
+        else {
+            ptr = gnrc_pktbuf_mark(snd_buf->pkt, remaining,
+                                   GNRC_NETTYPE_UNDEF);
+            if (ptr == NULL) {
+                DEBUG("ipv6_ext_frag: packet buffer full, canceling fragmentation\n");
+                gnrc_pktbuf_release(to_send);
+                _snd_buf_free(snd_buf);
+                return;
+            }
+            assert(snd_buf->pkt->next == ptr);  /* we just created it with mark */
+            snd_buf->pkt->next = snd_buf->pkt->next->next;
+            ptr->next = NULL;
+        }
+        last->next = ptr;
+        last = ptr;
+        remaining -= ptr->size;
+        snd_buf->offset += ptr->size;
+    }
+    assert(len != NULL);
+    /* adapt IPv6 header length field */
+    *len = byteorder_htons(gnrc_pkt_len(to_send->next->next));
+    msg.type = GNRC_IPV6_EXT_FRAG_SEND_FRAG;
+    msg.content.ptr = to_send;
+    msg_try_send(&msg, gnrc_ipv6_pid);
+    if (last_fragment) {
+        _snd_buf_del(snd_buf);
+    }
+    else {
+        msg.type = GNRC_IPV6_EXT_FRAG_SEND;
+        msg.content.ptr = snd_buf;
+        msg_try_send(&msg, gnrc_ipv6_pid);
+    }
+}
+
+static gnrc_ipv6_ext_frag_send_t *_snd_buf_alloc(void)
+{
+    for (unsigned i = 0; i < GNRC_IPV6_EXT_FRAG_SEND_SIZE; i++) {
+        gnrc_ipv6_ext_frag_send_t *snd_buf = &_snd_bufs[i];
+        if (snd_buf->pkt == NULL) {
+            return snd_buf;
+        }
+    }
+    return NULL;
+}
+
+static void _snd_buf_del(gnrc_ipv6_ext_frag_send_t *snd_buf)
+{
+    snd_buf->per_frag = NULL;
+    snd_buf->pkt = NULL;
+}
+
+static void _snd_buf_free(gnrc_ipv6_ext_frag_send_t *snd_buf)
+{
+    if (snd_buf->per_frag) {
+        gnrc_pktbuf_release(snd_buf->per_frag);
+    }
+    if (snd_buf->pkt) {
+        gnrc_pktbuf_release(snd_buf->pkt);
+    }
+    _snd_buf_del(snd_buf);
+}
+
+static gnrc_pktsnip_t *_determine_last_per_frag(gnrc_pktsnip_t *ptr)
+{
+    gnrc_pktsnip_t *last_per_frag = NULL;
+    unsigned nh = PROTNUM_RESERVED;
+
+    /* ignore NETIF header */
+    ptr = ptr->next;
+    while (ptr) {
+        switch (ptr->type) {
+            case GNRC_NETTYPE_IPV6: {
+                ipv6_hdr_t *hdr = ptr->data;
+                last_per_frag = ptr;
+                nh = hdr->nh;
+                break;
+            }
+            case GNRC_NETTYPE_IPV6_EXT: {
+                ipv6_ext_t *hdr = ptr->data;
+                switch (nh) {
+                    /* "[...] that is, all headers up to and including the
+                     * Routing header if present, else the Hop-by-Hop Options
+                     * header if present, [...]"
+                     * (IPv6 header comes before Hop-by-Hop Options comes before
+                     * Routing header, so an override to keep the quoted
+                     * priorities is ensured) */
```

Comment:
*No wonder you need fragmentation with so many headers…*

Reply:
That's why you have to have at least 1280 bytes PDU with IPv6 ;-).